### DISCUSSION TITLE
Negative values for BA using KDM

### DISCUSSION THREAD

**Discussion by caterincst** [dayoonkwon/BioAge#9] [Q&A] [answered]
Hello!

After calculating BA using the kdm_calc function, some of the estimated BA were negative and incredibly huge, and the biggest one was even up to 1000 years. And the correlation between BA and CA was weak (r=0.2). According to results of the analysis and BA formula of KDM, x-q in the numerator is the only possible way to make BA negative. Could you give some advice on why this might happen and how to deal with this problem? Thank you very much.
![图片](https://user-images.githubusercontent.com/100190890/155158305-12b26261-0579-43ab-93b7-5c49e9e86e0a.png)

**Comment by qu515** [comment]
We're encountering the same issue of negative kdm and phenoage values. The dataset we're trying to calculate biological age for contains NHANES data from 1999 to 2020. Age of the subjects ranges from 6 to 84, with pregnant individuals excluded. Biomarkers used are the same as those in the example projecting onto the CALERIE dataset; only subjects with at least 10 out of 12 biomarkers available are included. Among subjects in the lower age range (e.g. 12-19) we're seeing a particularly high number of negative biological age values. The following graphs show the biological age we calculated plotted against chronological age. Could you advise on why this may be and how best to address it? Thank you for your time in advance!

<img width="524" alt="Screen Shot 2023-01-18 at 3 03 14 PM" src="https://user-images.githubusercontent.com/122990988/213191672-c04a7304-3c19-41c0-a847-e4c45a8707ac.png">
<img width="511" alt="Screen Shot 2023-01-18 at 3 03 35 PM" src="https://user-images.githubusercontent.com/122990988/213191709-f6b147f2-47ed-4e21-808e-a6533e6a45b1.png">
<img width="524" alt="Screen Shot 2023-01-18 at 3 03 56 PM" src="https://user-images.githubusercontent.com/122990988/213191735-c58a30c3-c6d4-4b66-b1f5-d61476df6070.png">

**Reply by dayoonkwon** [reply]
In general, you can get implausible (including very large and/or negative) BA values. This tends to occur when you have non-normally distributed markers included in the algorithm being developed and/or in the test data. I would need to know a bit more about your application:

What is the sample you are using for training?
What is the sample you are using for testing?

Good first steps when assembling a biomarker panel for BA are to

1) Plot distributions of the markers in the training and test dataset
2) Create scatterplots of the association between chronological age and the biomarker in the training and test datasets

These two sets of figures will help you get a sense of where things might be going wrong.

Best,
Dayoon

**Reply by qu515** [reply]
Hi Dayoon, thank you very much for the quick answer! Training was done in **NHANES III** using the exact same code and biomarkers found in the README document.

hd = hd_nhanes(biomarkers=c("albumin","alp","lncrp","totchol","lncreat","hba1c","sbp","bun","uap","lymph","mcv","wbc"))
kdm = kdm_nhanes(biomarkers=c("albumin","alp","lncrp","totchol","lncreat","hba1c","sbp","bun","uap","lymph","mcv","wbc"))
phenoage = phenoage_nhanes(biomarkers=c("albumin_gL","alp","lncrp","totchol","lncreat_umol","hba1c","sbp","bun","uap","lymph","mcv","wbc"))

Then it was projected into **NHANES IV** (1999-2020). We did some cleaning of the dataset based on audiometry (since we're investigating biological age and hearing loss) and excluded pregnancy. 

We did plot the distributions of biomarkers in the test dataset. Some biomarkers are indeed not normally distributed. How should those be handled? Also, I notice that in the code for training (copied above) CRP and creatinine are log transformed. Is that because those two markers are not normally distributed in the NHANES III training set?

Thank you again for your help!

**Reply by dayoonkwon** [reply]
If you use your own cleaned NHANES dataset, you need to use the '(bioage)_calc' function instead of '(bioage)_nhanes'. This might be one reason for getting implausible values. You can handle non-normally distributed markers using transformations like what we did for CRP and creatinine. 

Best,
Dayoon

**Reply by qu515** [reply]
Hi again Dayoon, thank you for the advice. I did what you suggested and have a follow-up question: when using the kdm_calc and phenoage_calc functions for my own data, how should I handle the _fit_ argument? Is it mandatory to supply a _fit_ argument or could it be removed? 

For kdm, I implemented the following code without _fit_ or _s_ba2_ and was able to get better results than before (a lot fewer negative values).
```
kdm_fem = kdm_calc(data = dataset_all_clean %>%
                     filter (gender ==2),
                   biomarkers=c("albumin","alp","lncrp","totchol","lncreat","hba1c","sbp","bun","uap","lymph","mcv","wbc"))

kdm_male = kdm_calc(data = dataset_all_clean %>%
                      filter (gender ==1),
                    biomarkers=c("albumin","alp","lncrp","totchol","lncreat","hba1c","sbp","bun","uap","lymph","mcv","wbc"))
```

However, with phenoage_calc, when I tried implementing the following code without the _fit_ argument
```
phenoage_own = phenoage_calc(data = dataset_all_clean, 
                             biomarkers = c("albumin_gL","alp","lncrp","totchol","lncreat_umol","hba1c","sbp","bun","uap","lymph","mcv","wbc"))
```
I kept getting an error: "Error in Surv(time, status) : object 'status' not found." Do you have any advice on what might be causing the error? Thank you again for your time!

**Reply by dayoonkwon** [reply, marked as answer]
If you have already trained a kdm/phenoage model using a training dataset and want to apply this model to testing data, you should provide the `fit` argument using the trained algorithm. This allows the function to use the learned parameters for your calculations.
